### PR TITLE
Add a pair of hooks to DominanceInfo for MemToReg like pass

### PR DESCRIPTION
**Pull request by schweitzpgi**
These two hooks are to allow our implementation of the MemToReg like pass to transform the FIR dialect.

### CODE REVIEW

File: include/mlir/Analysis/Dominance.h [hunk 1]
```diff
@@ -90,6 +90,9 @@ class DominanceInfo : public detail::DominanceInfoBase</*IsPostDom=*/false> {
   bool properlyDominates(Block *a, Block *b) {
     return super::properlyDominates(a, b);
   }
+
+  llvm::DomTreeNodeBase<Block> *getNode(Block *a);
+  void updateDFSNumbers();
```
Comment (joker-eph) [inline include/mlir/Analysis/Dominance.h]: These are public APIs, can you provide documentation for them?

**Comment by schweitzpgi** [comment]
> Is this possible to get testing coverage for this in-tree?

Can you suggest the right way to test this interface? No test was provided here (although we exercise this by running mem2reg from the Fortran bits) because it wasn't totally clear what the best thing to do in the MLIR test environment.

File: include/mlir/Analysis/Dominance.h (hunk 1 above)
Comment (schweitzpgi) [reply include/mlir/Analysis/Dominance.h]: Will do.

File: include/mlir/Analysis/Dominance.h [hunk 2]
```diff
@@ -90,6 +90,12 @@ class DominanceInfo : public detail::DominanceInfoBase</*IsPostDom=*/false> {
   bool properlyDominates(Block *a, Block *b) {
     return super::properlyDominates(a, b);
   }
+
+  /// Return the dominanceInfos from the Region containing block A.
```
Comment (River707) [inline include/mlir/Analysis/Dominance.h]: dominanceInfos doesn't make sense here, can this change to 'dominance node'. This should also change the llvm::DomTreeNodeBase<Block> -> DominanceInfoNode

File: include/mlir/Analysis/Dominance.h [hunk 3]
```diff
@@ -90,6 +90,12 @@ class DominanceInfo : public detail::DominanceInfoBase</*IsPostDom=*/false> {
...
+  llvm::DomTreeNodeBase<Block> *getNode(Block *a);
+
+  /// Update the DFS numbers on all the dominanceInfos.
```
Comment (River707) [inline include/mlir/Analysis/Dominance.h]: dominanceInfos doesn't make sense here either, can we just say 'Updated the internal DFS numbers for the dominance nodes'

File: lib/Analysis/Dominance.cpp [hunk 4]
```diff
@@ -142,6 +142,18 @@ bool DominanceInfo::properlyDominates(Value *a, Operation *b) {
...
 }
 
+llvm::DomTreeNodeBase<Block> *DominanceInfo::getNode(Block *a) {
+  auto region = a->getParent();
+  assert(dominanceInfos.count(region) != 0);
+  return dominanceInfos[region]->getNode(a);
+}
+
+void DominanceInfo::updateDFSNumbers() {
+  for (auto &iter : dominanceInfos) {
```
Comment (River707) [inline lib/Analysis/Dominance.cpp]: Please drop the trivial braces.

File: include/mlir/Analysis/Dominance.h (hunk 2 above)
Comment (schweitzpgi) [reply include/mlir/Analysis/Dominance.h]: change pushed

File: include/mlir/Analysis/Dominance.h (hunk 3 above)
Comment (schweitzpgi) [reply include/mlir/Analysis/Dominance.h]: change pushed

File: lib/Analysis/Dominance.cpp (hunk 4 above)
Comment (schweitzpgi) [reply lib/Analysis/Dominance.cpp]: done

File: lib/Analysis/Dominance.cpp [hunk 5]
```diff
@@ -142,6 +142,17 @@ bool DominanceInfo::properlyDominates(Value *a, Operation *b) {
   return dominates(cast<BlockArgument>(a)->getOwner(), b->getBlock());
 }
 
+DominanceInfoNode *DominanceInfo::getNode(Block *a) {
+  auto region = a->getParent();
```
Comment (River707) [inline lib/Analysis/Dominance.cpp]: nit: Region *region or auto *region

**Comment by River707** [comment]
Orthogonally from this patch, how generic is the Mem2Reg pass for FIR? Would it be something that we could possibly contribute to upstream MLIR in the future(likely with some tweaks/generalizations)?

**Comment by schweitzpgi** [comment]
> Orthogonally from this patch, how generic is the Mem2Reg pass for FIR? Would it be something that we could possibly contribute to upstream MLIR in the future(likely with some tweaks/generalizations)?

The code was taken from LLVM and suitably modified to work with FIR Operations, etc.  I'm confident it can be made generic with some templates and work with any dialect. (Perhaps with the appropriate Interface; i.e. a Memory form that uses local temporaries.)

**Pull request closed without merging**